File: javascript/125.valid-palindrome.js
/*
 * @lc app=leetcode id=125 lang=javascript
 *
 * [125] Valid Palindrome
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
  let chars = s.match(/[a-zA-Z0-9]/g) ?? []
  if (chars.length === 0) return true
  return chars.join("").toLowerCase() === chars.reverse().join("").toLowerCase()
};
// @lc code=end

File: cpp/125.valid-palindrome.cpp
/*
 * @lc app=leetcode id=125 lang=cpp
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution
{
public:
    bool isPalindrome(string s)
    {
        // iterate pointers left and right
        // skip non-alphanumerics
        // O(n) time O(k) space
        int indexLow = 0, indexHigh = s.size() - 1;
        while (indexLow < indexHigh)
        {
            if (!isalnum(s[indexLow]))
            {
                indexLow++;
            }
            else if (!isalnum(s[indexHigh]))
            {
                indexHigh--;
            }
            else if (tolower(s[indexLow]) != tolower(s[indexHigh]))
            {
                return false;
            }
            else
            {
                indexLow++;
                indexHigh--;
            }
        }
        return true;
    }
};
// @lc code=end
